### ISSUE TITLE
GPDB 7 when use grouping sets like rollup, will first execute Motion to master then execute agg

### ISSUE THREAD

**Issue by 1Makers1** [greenplum-db/gpdb#9156]
### Greenplum version or build

```
postgres=# select version();
                                                                                            version
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 9.6beta4 (Greenplum Database 7.0.0-alpha.0 build dev) on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5), 64-bit compiled on Dec  3 2019 06:20:38
(1 row)
```
### OS version and uname -a
`Linux  3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
`
### Installation information ( pg_config )
$ pg_config
```
BINDIR = /usr/local/gpdb/bin
DOCDIR = /usr/local/gpdb/share/doc/postgresql
HTMLDIR = /usr/local/gpdb/share/doc/postgresql
INCLUDEDIR = /usr/local/gpdb/include
PKGINCLUDEDIR = /usr/local/gpdb/include/postgresql
INCLUDEDIR-SERVER = /usr/local/gpdb/include/postgresql/server
LIBDIR = /usr/local/gpdb/lib
PKGLIBDIR = /usr/local/gpdb/lib/postgresql
LOCALEDIR = /usr/local/gpdb/share/locale
MANDIR = /usr/local/gpdb/share/man
SHAREDIR = /usr/local/gpdb/share/postgresql
SYSCONFDIR = /usr/local/gpdb/etc/postgresql
PGXS = /usr/local/gpdb/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CONFIGURE = '--prefix=/home/test/dir_temp30/test/dist/gpdb-db' '--with-includes=/home/test/dir_temp30/test/orca/gp-xerces/dist/include:/home/test/dir_temp30/test/orca/gporca/dist/include:/home/test/dir_temp30/test/sigar/dist/include' '--with-libraries=/home/test/dir_temp30/test/orca/gporca/dist/lib:/home/test/dir_temp30/test/orca/gp-xerces/dist/lib:/home/test/dir_temp30/test/sigar/dist/lib' '--with-perl' '--with-libxml' '--enable-gpperfmon' '--enable-mapreduce' '--with-ldap' '--with-krb5' '--with-gssapi' '--with-openssl' '--with-python' '--enable-orca' '--enable-depend' '--enable-pxf' '--enable-gpfdist' '--disable-gpcloud' '--enable-debug' '--enable-orafce' '--enable-debug-extensions' '--disable-thread-safety'
CC = gcc
CPPFLAGS = -DFRONTEND -D_GNU_SOURCE -I/usr/include/libxml2 -I/home/test/dir_temp30/test/orca/gp-xerces/dist/include -I/home/test/dir_temp30/test/orca/gporca/dist/include -I/home/test/dir_temp30/test/sigar/dist/include
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-but-set-variable -Wno-format-truncation -g -ggdb -O3 -std=gnu99   -Werror=uninitialized -Werror=implicit-function-declaration
CFLAGS_SL = -fPIC
LDFLAGS = -L../../src/common -L/home/test/dir_temp30/test/orca/gporca/dist/lib -L/home/test/dir_temp30/test/orca/gp-xerces/dist/lib -L/home/test/dir_temp30/test/sigar/dist/lib -Wl,--as-needed -Wl,-rpath,'/home/test/dir_temp30/test/dist/gpdb-db/lib',--enable-new-dtags
LDFLAGS_EX =
LDFLAGS_SL =
LIBS = -lpgcommon -lpgport -lgpopt -lnaucrates -lgpdbcost -lgpos -lxerces-c -lbz2 -lxml2 -lrt -lssl -lcrypto -lyaml -lgssapi_krb5 -lzstd -lz -lreadline -lrt -lcrypt -ldl -lm  -lcurl
VERSION = PostgreSQL 9.6beta4
```
### Expected behavior
there some diff between GPDB6.0 and GPDB7.0
for example use rollup function
GPDB6.0 will first run agg in the segment then execute motion in the master
GPDB7.0  will first run motion to matser then execute agg in the master
the problem is in GPDB7.0 ,if the data number is large ,the master will spend a lot of time in agg.

### Actual behavior
```
postgres=# explain select  *
postgres-# from (select i_category
postgres(#             ,i_class
postgres(#             ,i_brand
postgres(#             ,i_product_name
postgres(#             ,d_year
postgres(#             ,d_qoy
postgres(#             ,d_moy
postgres(#             ,s_store_id
postgres(#             ,sumsales
postgres(#             ,rank() over (partition by i_category order by sumsales desc) rk
postgres(#       from (select i_category
postgres(#                   ,i_class
postgres(#                   ,i_brand
postgres(#                   ,i_product_name
postgres(#                   ,d_year
postgres(#                   ,d_qoy
postgres(#                   ,d_moy
postgres(#                   ,s_store_id
postgres(#                   ,sum(coalesce(ss_sales_price*ss_quantity,0)) sumsales
postgres(#             from store_sales
postgres(#                 ,date_dim
postgres(#                 ,store
postgres(#                 ,item
postgres(#        where  ss_sold_date_sk=d_date_sk
postgres(#           and ss_item_sk=i_item_sk
postgres(#           and ss_store_sk = s_store_sk
postgres(#           and d_month_seq between 1201 and 1201+11
postgres(#        group by  rollup(i_category, i_class, i_brand, i_product_name, d_year, d_qoy, d_moy,s_store_id))dw1) dw2
postgres-# where rk <= 100
postgres-# order by i_category
postgres-#         ,i_class
postgres-#         ,i_brand
postgres-#         ,i_product_name
postgres-#         ,d_year
postgres-#         ,d_qoy
postgres-#         ,d_moy
postgres-#         ,s_store_id
postgres-#         ,sumsales
postgres-#         ,rk
postgres-# limit 100;
                                                                                           QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=80159196.28..80159196.53 rows=100 width=135)
   ->  Sort  (cost=80159196.28..80204022.45 rows=17930469 width=135)
         Sort Key: dw2.i_category, dw2.i_class, dw2.i_brand, dw2.i_product_name, dw2.d_year, dw2.d_qoy, dw2.d_moy, dw2.s_store_id, dw2.sumsales, dw2.rk
         ->  Subquery Scan on dw2  (cost=77725686.01..79473906.67 rows=17930469 width=135)
               Filter: (dw2.rk <= 100)
               ->  WindowAgg  (cost=77725686.01..78801514.11 rows=53791406 width=135)
                     Partition By: dw1.i_category
                     Order By: dw1.sumsales
                     ->  Sort  (cost=77725686.01..77860164.52 rows=53791406 width=127)
                           Sort Key: dw1.i_category, dw1.sumsales DESC
                           ->  Subquery Scan on dw1  (cost=56089734.32..57968303.48 rows=53791406 width=127)
                                 ->  GroupAggregate  (cost=56089734.32..57430389.43 rows=53791406 width=127)
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name
                                       Group Key: item.i_category, item.i_class, item.i_brand
                                       Group Key: item.i_category, item.i_class
                                       Group Key: item.i_category
                                       Group Key: ()
                                       ->  Gather Motion 72:1  (slice5; segments: 72)  (cost=56089734.32..56390452.47 rows=13365252 width=109)
                                             Merge Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                             ->  Sort  (cost=56089734.32..56123147.45 rows=185629 width=109)
                                                   Sort Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                                   ->  Hash Join  (cost=19302.77..54507824.31 rows=185629 width=109)
                                                         Hash Cond: (store_sales.ss_item_sk = item.i_item_sk)
                                                         ->  Redistribute Motion 72:72  (slice3; segments: 72)  (cost=3761.77..54308511.10 rows=185629 width=63)
                                                               Hash Key: store_sales.ss_item_sk
                                                               ->  Hash Join  (cost=3761.77..54041206.08 rows=185629 width=63)
                                                                     Hash Cond: (store_sales.ss_store_sk = store.s_store_sk)
                                                                     ->  Hash Join  (cost=2106.49..53853420.32 rows=194363 width=54)
                                                                           Hash Cond: (store_sales.ss_sold_date_sk = date_dim.d_date_sk)
                                                                           ->  Seq Scan on store_sales  (cost=0.00..42911417.68 rows=39999833 width=38)
                                                                           ->  Hash  (cost=1772.02..1772.02 rows=372 width=32)
                                                                                 ->  Broadcast Motion 72:72  (slice1; segments: 72)  (cost=0.00..1772.02 rows=372 width=32)
                                                                                       ->  Seq Scan on date_dim  (cost=0.00..1500.73 rows=6 width=32)
                                                                                             Filter: ((d_month_seq >= 1201) AND (d_month_seq <= 1212))
                                                                     ->  Hash  (cost=753.48..753.48 rows=1002 width=25)
                                                                           ->  Broadcast Motion 72:72  (slice2; segments: 72)  (cost=0.00..753.48 rows=1002 width=25)
                                                                                 ->  Seq Scan on store  (cost=0.00..22.02 rows=14 width=25)
                                                         ->  Hash  (cost=11791.00..11791.00 rows=4167 width=62)
                                                               ->  Redistribute Motion 72:72  (slice4; segments: 72)  (cost=0.00..11791.00 rows=4167 width=62)
                                                                     Hash Key: item.i_item_sk
                                                                     ->  Seq Scan on item  (cost=0.00..5791.00 rows=4167 width=62)
 Optimizer: Postgres query optimizer
(46 rows)
```

### Step to reproduce the behavior
```
create table store_sales
(
    ss_sold_date_sk           bigint                       ,
    ss_sold_time_sk           bigint                       ,
    ss_item_sk                bigint                   ,
    ss_customer_sk            bigint                       ,
    ss_cdemo_sk               bigint                       ,
    ss_hdemo_sk               bigint                       ,
    ss_addr_sk                bigint                       ,
    ss_store_sk               bigint                       ,
    ss_promo_sk               bigint                       ,
    ss_ticket_number          bigint                   ,
    ss_quantity               bigint                       ,
    ss_wholesale_cost         decimal(7,2)                  ,
    ss_list_price             decimal(7,2)                  ,
    ss_sales_price            decimal(7,2)                  ,
    ss_ext_discount_amt       decimal(7,2)                  ,
    ss_ext_sales_price        decimal(7,2)                  ,
    ss_ext_wholesale_cost     decimal(7,2)                  ,
    ss_ext_list_price         decimal(7,2)                  ,
    ss_ext_tax                decimal(7,2)                  ,
    ss_coupon_amt             decimal(7,2)                  ,
    ss_net_paid               decimal(7,2)                  ,
    ss_net_paid_inc_tax       decimal(7,2)                  ,
    ss_net_profit             decimal(7,2)
) with (APPENDONLY=TRUE, ORIENTATION=column)  DISTRIBUTED RANDOMLY ;
create table date_dim
(
    d_date_sk                 bigint                   ,
    d_date_id                 varchar(16)                  ,
    d_date                    date                          ,
    d_month_seq               bigint                       ,
    d_week_seq                bigint                       ,
    d_quarter_seq             bigint                       ,
    d_year                    bigint                       ,
    d_dow                     bigint                       ,
    d_moy                     bigint                       ,
    d_dom                     bigint                       ,
    d_qoy                     bigint                       ,
    d_fy_year                 bigint                       ,
    d_fy_quarter_seq          bigint                       ,
    d_fy_week_seq             bigint                       ,
    d_day_name                varchar(9)                       ,
    d_quarter_name            varchar(6)                       ,
    d_holiday                 varchar(1)                       ,
    d_weekend                 varchar(1)                       ,
    d_following_holiday       varchar(1)                       ,
    d_first_dom               bigint                       ,
    d_last_dom                bigint                       ,
    d_same_day_ly             bigint                       ,
    d_same_day_lq             bigint                       ,
    d_current_day             varchar(1)                       ,
    d_current_week            varchar(1)                       ,
    d_current_month           varchar(1)                       ,
    d_current_quarter         varchar(1)                       ,
    d_current_year            varchar(1)
) with (APPENDONLY=TRUE, ORIENTATION=column)  DISTRIBUTED RANDOMLY ;
create table store
(
    s_store_sk                bigint                   ,
    s_store_id                varchar(16)                  ,
    s_rec_start_date          date                          ,
    s_rec_end_date            date                          ,
    s_closed_date_sk          bigint                       ,
    s_store_name              varchar(50)                   ,
    s_number_employees        bigint                       ,
    s_floor_space             bigint                       ,
    s_hours                   varchar(20)                      ,
    s_manager                 varchar(40)                   ,
    s_market_id               bigint                       ,
    s_geography_class         varchar(100)                  ,
    s_market_desc             varchar(100)                  ,
    s_market_manager          varchar(40)                   ,
    s_division_id             bigint                       ,
    s_division_name           varchar(50)                   ,
    s_company_id              bigint                       ,
    s_company_name            varchar(50)                   ,
    s_street_number           varchar(10)                   ,
    s_street_name             varchar(60)                   ,
    s_street_type             varchar(15)                      ,
    s_suite_number            varchar(10)                      ,
    s_city                    varchar(60)                   ,
    s_county                  varchar(30)                   ,
    s_state                   varchar(2)                       ,
    s_zip                     varchar(10)                      ,
    s_country                 varchar(20)                   ,
    s_gmt_offset              decimal(5,2)                  ,
    s_tax_precentage          decimal(5,2)
) with (APPENDONLY=TRUE, ORIENTATION=column)  DISTRIBUTED RANDOMLY ;
create table item
(
    i_item_sk                 bigint                   ,
    i_item_id                 varchar(16)                  ,
    i_rec_start_date          date                          ,
    i_rec_end_date            date                          ,
    i_item_desc               varchar(200)                  ,
    i_current_price           decimal(7,2)                  ,
    i_wholesale_cost          decimal(7,2)                  ,
    i_brand_id                bigint                       ,
    i_brand                   varchar(50)                      ,
    i_class_id                bigint                       ,
    i_class                   varchar(50)                      ,
    i_category_id             bigint                       ,
    i_category                varchar(50)                      ,
    i_manufact_id             bigint                       ,
    i_manufact                varchar(50)                      ,
    i_size                    varchar(20)                      ,
    i_formulation             varchar(20)                      ,
    i_color                   varchar(20)                      ,
    i_units                   varchar(10)                      ,
    i_container               varchar(10)                      ,
    i_manager_id              bigint                       ,
    i_product_name            varchar(50)
) with (APPENDONLY=TRUE, ORIENTATION=column)  DISTRIBUTED RANDOMLY ;
```

```
postgres=# show optimizer;
 optimizer
-----------
 on

(1 row)
```

```
postgres=# explain select  *
from (select i_category
            ,i_class
            ,i_brand
            ,i_product_name
            ,d_year
            ,d_qoy
            ,d_moy
            ,s_store_id
            ,sumsales
            ,rank() over (partition by i_category order by sumsales desc) rk
      from (select i_category
                  ,i_class
                  ,i_brand
                  ,i_product_name
                  ,d_year
                  ,d_qoy
                  ,d_moy
                  ,s_store_id
                  ,sum(coalesce(ss_sales_price*ss_quantity,0)) sumsales
            from store_sales
                ,date_dim
                ,store
                ,item
       where  ss_sold_date_sk=d_date_sk
          and ss_item_sk=i_item_sk
          and ss_store_sk = s_store_sk
          and d_month_seq between 1201 and 1201+11
       group by  rollup(i_category, i_class, i_brand, i_product_name, d_year, d_qoy, d_moy,s_store_id))dw1) dw2
where rk <= 100
order by i_category
        ,i_class
        ,i_brand
        ,i_product_name
        ,d_year
        ,d_qoy
        ,d_moy
        ,s_store_id
        ,sumsales
        ,rk
limit 100;
                                                                                           QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=80159196.28..80159196.53 rows=100 width=135)
   ->  Sort  (cost=80159196.28..80204022.45 rows=17930469 width=135)
         Sort Key: dw2.i_category, dw2.i_class, dw2.i_brand, dw2.i_product_name, dw2.d_year, dw2.d_qoy, dw2.d_moy, dw2.s_store_id, dw2.sumsales, dw2.rk
         ->  Subquery Scan on dw2  (cost=77725686.01..79473906.67 rows=17930469 width=135)
               Filter: (dw2.rk <= 100)
               ->  WindowAgg  (cost=77725686.01..78801514.11 rows=53791406 width=135)
                     Partition By: dw1.i_category
                     Order By: dw1.sumsales
                     ->  Sort  (cost=77725686.01..77860164.52 rows=53791406 width=127)
                           Sort Key: dw1.i_category, dw1.sumsales DESC
                           ->  Subquery Scan on dw1  (cost=56089734.32..57968303.48 rows=53791406 width=127)
                                 ->  GroupAggregate  (cost=56089734.32..57430389.43 rows=53791406 width=127)
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name
                                       Group Key: item.i_category, item.i_class, item.i_brand
                                       Group Key: item.i_category, item.i_class
                                       Group Key: item.i_category
                                       Group Key: ()
                                       ->  Gather Motion 72:1  (slice5; segments: 72)  (cost=56089734.32..56390452.47 rows=13365252 width=109)
                                             Merge Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                             ->  Sort  (cost=56089734.32..56123147.45 rows=185629 width=109)
                                                   Sort Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                                   ->  Hash Join  (cost=19302.77..54507824.31 rows=185629 width=109)
                                                         Hash Cond: (store_sales.ss_item_sk = item.i_item_sk)
                                                         ->  Redistribute Motion 72:72  (slice3; segments: 72)  (cost=3761.77..54308511.10 rows=185629 width=63)
                                                               Hash Key: store_sales.ss_item_sk
                                                               ->  Hash Join  (cost=3761.77..54041206.08 rows=185629 width=63)
                                                                     Hash Cond: (store_sales.ss_store_sk = store.s_store_sk)
                                                                     ->  Hash Join  (cost=2106.49..53853420.32 rows=194363 width=54)
                                                                           Hash Cond: (store_sales.ss_sold_date_sk = date_dim.d_date_sk)
                                                                           ->  Seq Scan on store_sales  (cost=0.00..42911417.68 rows=39999833 width=38)
                                                                           ->  Hash  (cost=1772.02..1772.02 rows=372 width=32)
                                                                                 ->  Broadcast Motion 72:72  (slice1; segments: 72)  (cost=0.00..1772.02 rows=372 width=32)
                                                                                       ->  Seq Scan on date_dim  (cost=0.00..1500.73 rows=6 width=32)
                                                                                             Filter: ((d_month_seq >= 1201) AND (d_month_seq <= 1212))
                                                                     ->  Hash  (cost=753.48..753.48 rows=1002 width=25)
                                                                           ->  Broadcast Motion 72:72  (slice2; segments: 72)  (cost=0.00..753.48 rows=1002 width=25)
                                                                                 ->  Seq Scan on store  (cost=0.00..22.02 rows=14 width=25)
                                                         ->  Hash  (cost=11791.00..11791.00 rows=4167 width=62)
                                                               ->  Redistribute Motion 72:72  (slice4; segments: 72)  (cost=0.00..11791.00 rows=4167 width=62)
                                                                     Hash Key: item.i_item_sk
                                                                     ->  Seq Scan on item  (cost=0.00..5791.00 rows=4167 width=62)
 Optimizer: Postgres query optimizer
(46 rows)
```

this use tpcds test then find this problem ， can  use the tpcds to inspect more problem

**Comment by Eulerizeit** [comment]
This is a known issue. When we merged in the upperpathification work in 9.6 we greatly simplified our GROUPING SETs code. 

We are currently reimplementing the feature in a more intelligent and performant way. Please follow along and test the PR when it is posted shortly.

**Comment by 1Makers1** [comment]
Thanks for your reply, when this problem is solved, I will try again.  Ask offensively, when will it be resolved,or is there any way to circumvent this problem without changing sql，thank you reply.

**Comment by Eulerizeit** [comment]
@1Makers1 here is the PR.  https://github.com/greenplum-db/gpdb/pull/9219 any help with review or testing would be appreciated.

**Comment by 1Makers1** [comment]
i merge  [#9219](https://github.com/greenplum-db/gpdb/pull/9219 ) this patch ,and have
a test ,but it doesn't seem to work for this issue.

**Comment by guofengrichard** [comment]
> i merge [#9219](https://github.com/greenplum-db/gpdb/pull/9219) this patch ,and have
> a test ,but it doesn't seem to work for this issue.

I debugged your case and found that the two-stage path was generated but defeated by the one-stage path in add_path(). 

Would you please confirm that the two-stage path indeed has less execution time than the one-stage path? If that's the case, something goes wrong with the cost calculation.

**Comment by 1Makers1** [comment]
For the code which i merge the patch, I used a larger amount of data to test. it seems to
spend a lot of time on the master. If they run on the segment instances, they can perform operations in parallel, which costs may be more expensive, but because it is multiple parallel operations, it may take less time.

```
postgres=# \d+
                                     List of relations
 Schema |      Name       | Type  |  Owner  |       Storage        |  Size   | Description
--------+-----------------+-------+---------+----------------------+---------+-------------
 public | date_dim        | table | ssadmin | append only columnar | 16 MB   |
 public | ext_date_dim    | table | ssadmin | external             | 0 bytes |
 public | ext_item        | table | ssadmin | external             | 0 bytes |
 public | ext_store       | table | ssadmin | external             | 0 bytes |
 public | ext_store_sales | table | ssadmin | external             | 0 bytes |
 public | item            | table | ssadmin | append only columnar | 90 MB   |
 public | store           | table | ssadmin | append only columnar | 3560 kB |
 public | store_sales     | table | ssadmin | append only columnar | 172 GB  |
 public | time_dim        | table | ssadmin | append only columnar | 544 kB  |

```

```
postgres=# explain analyze select  *
from (select i_category
            ,i_class
            ,i_brand
            ,i_product_name
            ,d_year
            ,d_qoy
            ,d_moy
            ,s_store_id
            ,sumsales
            ,rank() over (partition by i_category order by sumsales desc) rk
      from (select i_category
                  ,i_class
                  ,i_brand
                  ,i_product_name
                  ,d_year
                  ,d_qoy
                  ,d_moy
                  ,s_store_id
                  ,sum(coalesce(ss_sales_price*ss_quantity,0)) sumsales
            from store_sales
                ,date_dim
                ,store
                ,item
       where  ss_sold_date_sk=d_date_sk
          and ss_item_sk=i_item_sk
          and ss_store_sk = s_store_sk
          and d_month_seq between 1220 and 1220+11
       group by  rollup(i_category, i_class, i_brand, i_product_name, d_year, d_qoy, d_moy,s_store_id))dw1) dw2
where rk <= 100
order by i_category
        ,i_class
        ,i_brand
        ,i_product_name
        ,d_year
        ,d_qoy
        ,d_moy
        ,s_store_id
        ,sumsales
        ,rk
limit 100;
                                                                                                         QUERY PLAN

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-------------
 Limit  (cost=31362792.43..31362792.68 rows=100 width=135) (actual time=2752427.406..2752427.436 rows=100 loops=1)
   ->  Sort  (cost=31362792.43..31378184.97 rows=6157014 width=135) (actual time=2752427.403..2752427.408 rows=100 loops=1)
         Sort Key: dw2.i_category, dw2.i_class, dw2.i_brand, dw2.i_product_name, dw2.d_year, dw2.d_qoy, dw2.d_moy, dw2.s_store_id, dw2.sumsales, dw2.rk
         Sort Method:  top-N heapsort  Memory: 57kB
         ->  Subquery Scan on dw2  (cost=30527166.94..31127475.79 rows=6157014 width=135) (actual time=2516734.701..2752426.150 rows=1100 loops=1)
               Filter: (dw2.rk <= 100)
               Rows Removed by Filter: 189384453
               ->  WindowAgg  (cost=30527166.94..30896587.77 rows=18471042 width=135) (actual time=2516734.688..2740292.476 rows=189385553 loops=1)
                     Partition By: dw1.i_category
                     Order By: dw1.sumsales
                     ->  Sort  (cost=30527166.94..30573344.54 rows=18471042 width=127) (actual time=2516734.647..2625528.304 rows=189385553 loops=1)
                           Sort Key: dw1.i_category, dw1.sumsales DESC
                           Sort Method:  external merge  Disk: 23640352kB
                           ->  Subquery Scan on dw1  (cost=23242844.86..23885203.55 rows=18471042 width=127) (actual time=211042.846..1362189.437 rows=189385553 loops=1)
                                 ->  GroupAggregate  (cost=23242844.86..23700493.14 rows=18471042 width=127) (actual time=211042.844..1341156.493 rows=189385553 loops=1)
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name
                                       Group Key: item.i_category, item.i_class, item.i_brand
                                       Group Key: item.i_category, item.i_class
                                       Group Key: item.i_category
                                       Group Key: ()
                                       ->  Gather Motion 16:1  (slice5; segments: 16)  (cost=23242844.86..23344886.97 rows=4535206 width=109) (actual time=211042.763..1001061.274 rows=214873679 loops=1)
                                             Merge Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                             ->  Sort  (cost=23242844.86..23254182.87 rows=283451 width=109) (actual time=210885.646..302748.167 rows=13627193 loops=1)
                                                   Sort Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                                   Sort Method:  external merge  Disk: 28562336kB
                                                   ->  Hash Join  (cost=17545.73..21772031.38 rows=283451 width=109) (actual time=39.474..35087.702 rows=13627193 loops=1)
                                                         Hash Cond: (store_sales.ss_store_sk = store.s_store_sk)
                                                         Extra Text: (seg7)   Hash chain length 1.0 avg, 2 max, using 997 of 65536 buckets.
                                                         ->  Hash Join  (cost=17154.97..21708548.28 rows=295672 width=100) (actual time=38.259..29481.652 rows=13955708 loops=1)
                                                               Hash Cond: (store_sales.ss_item_sk = item.i_item_sk)
                                                               Extra Text: (seg7)   Hash chain length 1.1 avg, 4 max, using 16376 of 65536 buckets.
                                                               ->  Redistribute Motion 16:16  (slice2; segments: 16)  (cost=1614.97..21627960.57 rows=295672 width=54) (actual time=0.023..18317.746 rows=13955
708 loops=1)
                                                                     Hash Key: store_sales.ss_item_sk
                                                                     ->  Hash Join  (cost=1614.97..21533345.72 rows=295672 width=54) (actual time=6.666..75249.713 rows=13758386 loops=1)
                                                                           Hash Cond: (store_sales.ss_sold_date_sk = date_dim.d_date_sk)
                                                                           Extra Text: (seg3)   Hash chain length 1.0 avg, 2 max, using 364 of 65536 buckets.
                                                                           ->  Seq Scan on store_sales  (cost=0.00..17164467.96 rows=71999256 width=38) (actual time=2.227..55759.101 rows=72012982 loops=1)
                                                                           ->  Hash  (cost=1552.14..1552.14 rows=315 width=32) (actual time=2.937..2.937 rows=365 loops=1)
                                                                                 ->  Broadcast Motion 16:16  (slice1; segments: 16)  (cost=0.00..1552.14 rows=315 width=32) (actual time=0.020..2.809 rows=365
loops=1)
                                                                                       ->  Seq Scan on date_dim  (cost=0.00..1498.73 rows=20 width=32) (actual time=1.580..2.281 rows=32 loops=1)
                                                                                             Filter: ((d_month_seq >= 1220) AND (d_month_seq <= 1231))
                                                               ->  Hash  (cost=11790.00..11790.00 rows=18750 width=62) (actual time=37.950..37.950 rows=18969 loops=1)
                                                                     ->  Redistribute Motion 16:16  (slice3; segments: 16)  (cost=0.00..11790.00 rows=18750 width=62) (actual time=0.038..31.319 rows=18969 loo
ps=1)
                                                                           Hash Key: item.i_item_sk
                                                                           ->  Seq Scan on item  (cost=0.00..5790.00 rows=18750 width=62) (actual time=1.006..8.672 rows=18984 loops=1)
                                                         ->  Hash  (cost=190.36..190.36 rows=1002 width=25) (actual time=2.340..2.340 rows=1002 loops=1)
                                                               ->  Broadcast Motion 16:16  (slice4; segments: 16)  (cost=0.00..190.36 rows=1002 width=25) (actual time=0.057..1.838 rows=1002 loops=1)
                                                                     ->  Seq Scan on store  (cost=0.00..20.02 rows=63 width=25) (actual time=1.036..1.083 rows=76 loops=1)
 Planning time: 3.040 ms
 * (slice0)    Executor memory: 131077K bytes.  Work_mem: 129925K bytes max, 41680343K bytes wanted.
   (slice1)    Executor memory: 592K bytes avg x 16 workers, 592K bytes max (seg0).
   (slice2)    Executor memory: 2969K bytes avg x 16 workers, 2969K bytes max (seg0).  Work_mem: 532K bytes max.
   (slice3)    Executor memory: 592K bytes avg x 16 workers, 592K bytes max (seg0).
   (slice4)    Executor memory: 300K bytes avg x 16 workers, 300K bytes max (seg0).
 * (slice5)    Executor memory: 47886K bytes avg x 16 workers, 47920K bytes max (seg11).  Work_mem: 44724K bytes max, 5430915K bytes wanted.
 Memory used:  128000kB
 Memory wanted:  291763994kB
 Optimizer: Postgres query optimizer
 Execution time: 2752491.350 ms
(63 rows)
```
Can you give me more suggestions and ideas, so that I can do more thing and give more information

**Comment by 1Makers1** [comment]
the sql in gpdb7.0 is much slower  than the sql in gpdb6.0  when the data is large.
attach gpdb6.0 explain.

```
postgres=# explain select  *
from (select i_category
            ,i_class
            ,i_brand
            ,i_product_name
            ,d_year
            ,d_qoy
            ,d_moy
            ,s_store_id
            ,sumsales
            ,rank() over (partition by i_category order by sumsales desc) rk
      from (select i_category
                  ,i_class
                  ,i_brand
                  ,i_product_name
                  ,d_year
                  ,d_qoy
                  ,d_moy
                  ,s_store_id
                  ,sum(coalesce(ss_sales_price*ss_quantity,0)) sumsales
            from store_sales
                ,date_dim
                ,store
                ,item
       where  ss_sold_date_sk=d_date_sk
          and ss_item_sk=i_item_sk
          and ss_store_sk = s_store_sk
          and d_month_seq between 1201 and 1201+11
       group by  rollup(i_category, i_class, i_brand, i_product_name, d_year, d_qoy, d_moy,s_store_id))dw1) dw2
where rk <= 100
order by i_category
        ,i_class
        ,i_brand
        ,i_product_name
        ,d_year
        ,d_qoy
        ,d_moy
        ,s_store_id
        ,sumsales
        ,rk
limit 100;
                                                                                                                                                                                    QUERY PLAN

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=9840.03..9842.28 rows=100 width=586)
   ->  Gather Motion 3:1  (slice7; segments: 3)  (cost=9840.03..9842.28 rows=100 width=586)
         Merge Key: dw2.i_category, dw2.i_class, dw2.i_brand, dw2.i_product_name, dw2.d_year, dw2.d_qoy, dw2.d_moy, dw2.s_store_id, dw2.sumsales, dw2.rk
         ->  Limit  (cost=9840.03..9840.28 rows=34 width=586)
               ->  Sort  (cost=9840.03..9864.19 rows=3222 width=586)
                     Sort Key: dw2.i_category, dw2.i_class, dw2.i_brand, dw2.i_product_name, dw2.d_year, dw2.d_qoy, dw2.d_moy, dw2.s_store_id, dw2.sumsales, dw2.rk
                     ->  Subquery Scan on dw2  (cost=8528.21..9470.61 rows=3222 width=586)
                           Filter: (dw2.rk <= 100)
                           ->  WindowAgg  (cost=8528.21..9108.15 rows=9666 width=578)
                                 Partition By: dw1.i_category
                                 Order By: dw1.sumsales
                                 ->  Sort  (cost=8528.21..8600.70 rows=9666 width=578)
                                       Sort Key: dw1.i_category, dw1.sumsales
                                       ->  Redistribute Motion 3:3  (slice6; segments: 3)  (cost=5146.63..6379.01 rows=9666 width=578)
                                             Hash Key: dw1.i_category
                                             ->  Subquery Scan on dw1  (cost=5146.63..5799.07 rows=9666 width=578)
                                                   ->  HashAggregate  (cost=5146.63..5509.10 rows=9666 width=578)
                                                         Group Key: partial_aggregation.i_category, partial_aggregation.i_class, partial_aggregation.i_brand, partial_aggregation.i_product_name, partial_aggregation.d_year, partial_aggrega
tion.d_qoy, partial_aggregation.d_moy, partial_aggregation.s_store_id, partial_aggregation."grouping", partial_aggregation."group_id"
                                                         ->  Subquery Scan on partial_aggregation  (cost=3647.18..5117.19 rows=302 width=590)
                                                               ->  Redistribute Motion 3:3  (slice5; segments: 3)  (cost=3647.18..5108.13 rows=302 width=590)
                                                                     Hash Key: "rollup".i_category, "rollup".i_class, "rollup".i_brand, "rollup".i_product_name, "rollup".d_year, "rollup".d_qoy, "rollup".d_moy, "rollup".s_store_id
                                                                     ->  GroupAggregate  (cost=3647.18..5090.01 rows=302 width=590)
                                                                           Group Key: "rollup"."grouping", "rollup"."group_id", "rollup".i_brand, "rollup".i_product_name, "rollup".d_year, "rollup".d_qoy, "rollup".d_moy, "rollup".s_store_
id, "rollup".i_category, "rollup".i_class
                                                                           ->  Subquery Scan on "rollup"  (cost=3647.18..5034.52 rows=453 width=590)
                                                                                 ->  GroupAggregate  (cost=3647.18..5020.93 rows=453 width=590)
                                                                                       Group Key: rollup_1.i_category, rollup_1."grouping", rollup_1."group_id", rollup_1.i_product_name, rollup_1.d_year, rollup_1.d_qoy, rollup_1.d_moy, ro
llup_1.s_store_id, rollup_1.i_class, rollup_1.i_brand
                                                                                       ->  Subquery Scan on rollup_1  (cost=3647.18..4945.05 rows=604 width=590)
                                                                                             ->  GroupAggregate  (cost=3647.18..4926.93 rows=604 width=590)
                                                                                                   Group Key: rollup_2.i_category, rollup_2.i_class, rollup_2."grouping", rollup_2."group_id", rollup_2.d_year, rollup_2.d_qoy, rollup_2.d_mo
y, rollup_2.s_store_id, rollup_2.i_brand, rollup_2.i_product_name
                                                                                                   ->  Subquery Scan on rollup_2  (cost=3647.18..4830.67 rows=755 width=590)
                                                                                                         ->  GroupAggregate  (cost=3647.18..4808.02 rows=755 width=590)
                                                                                                               Group Key: rollup_3.i_category, rollup_3.i_class, rollup_3.i_brand, rollup_3."grouping", rollup_3."group_id", rollup_3.d_qoy,
rollup_3.d_moy, rollup_3.s_store_id, rollup_3.i_product_name, rollup_3.d_year
                                                                                                               ->  Subquery Scan on rollup_3  (cost=3647.18..4691.37 rows=906 width=590)
                                                                                                                     ->  GroupAggregate  (cost=3647.18..4664.19 rows=906 width=590)
                                                                                                                           Group Key: rollup_4.i_category, rollup_4.i_class, rollup_4.i_brand, rollup_4.i_product_name, rollup_4."grouping",
rollup_4."group_id", rollup_4.d_moy, rollup_4.s_store_id, rollup_4.d_year, rollup_4.d_qoy
                                                                                                                           ->  Subquery Scan on rollup_4  (cost=3647.18..4527.16 rows=1057 width=590)
                                                                                                                                 ->  GroupAggregate  (cost=3647.18..4495.45 rows=1057 width=590)
                                                                                                                                       Group Key: rollup_5.i_category, rollup_5.i_class, rollup_5.i_brand, rollup_5.i_product_name, rollup_5.
d_year, rollup_5."grouping", rollup_5."group_id", rollup_5.s_store_id, rollup_5.d_qoy, rollup_5.d_moy
                                                                                                                                       ->  Subquery Scan on rollup_5  (cost=3647.18..4338.03 rows=1208 width=590)
                                                                                                                                             ->  GroupAggregate  (cost=3647.18..4301.79 rows=1208 width=590)
                                                                                                                                                   Group Key: rollup_6.i_category, rollup_6.i_class, rollup_6.i_brand, rollup_6.i_product_nam
e, rollup_6.d_year, rollup_6.d_qoy, rollup_6."grouping", rollup_6."group_id", rollup_6.d_moy, rollup_6.s_store_id
                                                                                                                                                   ->  Subquery Scan on rollup_6  (cost=3647.18..4123.99 rows=1359 width=590)
                                                                                                                                                         ->  GroupAggregate  (cost=3647.18..4083.22 rows=1359 width=590)
                                                                                                                                                               Group Key: rollup_7.i_category, rollup_7.i_class, rollup_7.i_brand, rollup_7.i
_product_name, rollup_7.d_year, rollup_7.d_qoy, rollup_7.d_moy, rollup_7."grouping", rollup_7."group_id", rollup_7.s_store_id
                                                                                                                                                               ->  Subquery Scan on rollup_7  (cost=3647.18..3885.03 rows=1510 width=590)
                                                                                                                                                                     ->  GroupAggregate  (cost=3647.18..3839.73 rows=1510 width=590)
                                                                                                                                                                           Group Key: item.i_category, item.i_class, item.i_brand, item.i_pro
duct_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                                                                                                                                                           ->  Sort  (cost=3647.18..3658.51 rows=1511 width=568)
                                                                                                                                                                                 Sort Key: item.i_category, item.i_class, item.i_brand, item.
i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                                                                                                                                                                 ->  Hash Join  (cost=579.69..3372.03 rows=1511 width=568)
                                                                                                                                                                                       Hash Cond: (store_sales.ss_item_sk = item.i_item_sk)
                                                                                                                                                                                       ->  Broadcast Motion 3:3  (slice4; segments: 3)  (cost
=427.94..1461.14 rows=1970 width=104)
                                                                                                                                                                                             ->  Hash Join  (cost=427.94..1382.35 rows=657 wi
dth=104)
                                                                                                                                                                                                   Hash Cond: (store_sales.ss_store_sk = stor
e.s_store_sk)
                                                                                                                                                                                                   ->  Redistribute Motion 3:3  (slice2; segm
ents: 3)  (cost=251.44..937.17 rows=365 width=62)
                                                                                                                                                                                                         Hash Key: store_sales.ss_store_sk
                                                                                                                                                                                                         ->  Hash Join  (cost=251.44..915.28
rows=365 width=62)
                                                                                                                                                                                                               Hash Cond: (store_sales.ss_sol
d_date_sk = date_dim.d_date_sk)
                                                                                                                                                                                                               ->  Seq Scan on store_sales  (
cost=0.00..214.00 rows=3800 width=46)
                                                                                                                                                                                                               ->  Hash  (cost=247.84..247.84
 rows=96 width=32)
                                                                                                                                                                                                                     ->  Broadcast Motion 3:3
  (slice1; segments: 3)  (cost=0.00..247.84 rows=96 width=32)
                                                                                                                                                                                                                           ->  Seq Scan on da
te_dim  (cost=0.00..244.00 rows=32 width=32)
                                                                                                                                                                                                                                 Filter: ((d_
month_seq >= 1201) AND (d_month_seq <= 1212))
                                                                                                                                                                                                   ->  Hash  (cost=154.00..154.00 rows=600 wi
dth=58)
                                                                                                                                                                                                         ->  Redistribute Motion 3:3  (slice3
; segments: 3)  (cost=0.00..154.00 rows=600 width=58)
                                                                                                                                                                                                               Hash Key: store.s_store_sk
                                                                                                                                                                                                               ->  Seq Scan on store  (cost=0
.00..118.00 rows=600 width=58)
                                                                                                                                                                                       ->  Hash  (cost=123.00..123.00 rows=767 width=480)
                                                                                                                                                                                             ->  Seq Scan on item  (cost=0.00..123.00 rows=76
7 width=480)
 Optimizer: Postgres query optimizer
(70 rows)

```

**Comment by hlinnaka** [comment]
I believe https://github.com/greenplum-db/gpdb/pull/10759 will address this, too.

**Comment by kainwen** [comment]
> I believe #10759 will address this, too. 

Hi @hlinnaka 

Actually, PR https://github.com/greenplum-db/gpdb/pull/10759 is merged into master branch. But we still do not handle rollup well, a very simple query like:

```sql
gpadmin=# select version();
                                                                                                              version

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------
 PostgreSQL 12beta2 (Greenplum Database 7.0.0-alpha.0+dev.14042.g5028054e9a build dev) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, 64-bit compil
ed on Nov  5 2020 23:16:54 (with assert checking)
(1 row)

gpadmin=# create table t(a int, b int, c int);
psql: NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
gpadmin=# explain select  a, b, sum(c) from t group by rollup(a, b);
                                        QUERY PLAN
------------------------------------------------------------------------------------------
 MixedAggregate  (cost=0.00..2588.74 rows=8791 width=16)
   Hash Key: a, b
   Hash Key: a
   Group Key: ()
   Planned Partitions: 4
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1332.33 rows=77900 width=12)
         ->  Seq Scan on t  (cost=0.00..293.67 rows=25967 width=12)
 Optimizer: Postgres query optimizer
(8 rows)
```

It still can only gather to QD then agg.

**Comment by kainwen** [comment]
Plan for 6X:

```sql
psql (9.4.24)
Type "help" for help.

gpadmin=# create table t(a int, b int, c int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
gpadmin=# explain select a, b, sum(c) from t group by rollup(a, b);
                                                     QUERY PLAN
--------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=8634.64..8887.81 rows=25317 width=28)
   ->  HashAggregate  (cost=8634.64..8887.81 rows=8439 width=28)
         Group Key: "rollup".a, "rollup".b, "rollup"."grouping", "rollup"."group_id"
         ->  Subquery Scan on "rollup"  (cost=7208.12..8585.96 rows=1299 width=28)
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=7208.12..8547.01 rows=1299 width=28)
                     Hash Key: rollup_1.a, rollup_1.b, (Grouping), group_id()
                     ->  GroupAggregate  (cost=7208.12..8469.11 rows=1299 width=28)
                           Group Key: rollup_1."grouping", rollup_1."group_id", rollup_1.a, rollup_1.b
                           ->  Subquery Scan on rollup_1  (cost=7208.12..8357.13 rows=1948 width=28)
                                 ->  GroupAggregate  (cost=7208.12..8298.71 rows=1948 width=28)
                                       Group Key: rollup_2.a, rollup_2."grouping", rollup_2."group_id", rollup_2.b
                                       ->  Subquery Scan on rollup_2  (cost=7208.12..8142.92 rows=2597 width=28)
                                             ->  GroupAggregate  (cost=7208.12..8065.02 rows=2597 width=28)
                                                   Group Key: t.a, t.b
                                                   ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
                                                         Sort Key: t.a, t.b
                                                         ->  Seq Scan on t  (cost=0.00..879.00 rows=25967 width=12)
 Optimizer: Postgres query optimizer
(18 rows)
```

**Comment by kainwen** [comment]
With my pr https://github.com/greenplum-db/gpdb/pull/11151
and some more correct of groups estimation of rollups, the plan in my local env for 1TB data is two stage
distributed agg.

But it still take huge time to finish (compared to 6X). I have some guess for this, once verified, I will update on this.

**Comment by kainwen** [comment]
See gpdb-dev's thread: https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/zCQ1wheKDu4

This issue will be fixed after pr #11151 goes in.

And we will get much faster after we enable parallel related stuff in gpdb.

**Comment by kainwen** [comment]
Fixed by https://github.com/greenplum-db/gpdb/pull/11151.

I test in my local env, this query (TPC-DS Q167, 1TB data, 4host*8segs), planner now generates two stage distributed agg for it.
